“@AssertFalse” Not Supported “Boolean”

When I use `Boolean+@ AssertFalse`, the compilation error is:  This attribute is not written to the schema, transform to LiteralBooleanSchema before writing

```
@Data
public class FormValidatorSaveVO implements Serializable {

    @AssertFalse(message = "必须是false")
    private Boolean assertFalse2;
    @AssertTrue(message = "必须是true")
    private Boolean assertTrue2;

}

```

error: 
```
[ERROR] Failed to execute goal sh.ivan:java-to-zod-maven-plugin:0.6.1:generate (generate) on project lamp-boot-server-datasource: Execution generate of goal sh.ivan:java-to-zod-maven-plugin:0.6.1:generate failed: This attribute is not written to the schema, transform to LiteralBooleanSchema before writing -> [Help 1]
```

--- a/java-to-zod-core/src/main/java/sh/ivan/zod/JavaToZodConverter.java
+++ b/java-to-zod-core/src/main/java/sh/ivan/zod/JavaToZodConverter.java
@@ -101,12 +101,10 @@ public class JavaToZodConverter {
 
     private Schema getBooleanSchema(Set<Attribute> attributes) {
         var equalsBooleanAttribute = attributes.stream()
-                .filter(attribute -> attribute instanceof EqualsBooleanAttribute)
+                .filter(attribute -> attribute.isOfType(EqualsBooleanAttribute.class))
                 .findAny();
         if (equalsBooleanAttribute.isPresent()) {
-            return new LiteralBooleanSchema(
-                    ((EqualsBooleanAttribute) equalsBooleanAttribute.get()).isValue(),
-                    Sets.difference(attributes, Set.of(equalsBooleanAttribute.get())));
+            return LiteralBooleanSchema.fromAttributes(equalsBooleanAttribute.get(), attributes);
         }
         return new BooleanSchema(attributes);
     }

--- a/java-to-zod-core/src/main/java/sh/ivan/zod/schema/LiteralBooleanSchema.java
+++ b/java-to-zod-core/src/main/java/sh/ivan/zod/schema/LiteralBooleanSchema.java
@@ -1,24 +1,49 @@
 package sh.ivan.zod.schema;
 
+import com.google.common.collect.Sets;
 import java.util.Set;
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.ToString;
 import sh.ivan.zod.schema.attribute.Attribute;
+import sh.ivan.zod.schema.attribute.AttributeWithMessage;
+import sh.ivan.zod.schema.attribute.EqualsBooleanAttribute;
 
 @Getter
 @ToString(callSuper = true)
 @EqualsAndHashCode(callSuper = true)
 public class LiteralBooleanSchema extends Schema {
     private final boolean value;
+    private final String message;
 
-    public LiteralBooleanSchema(boolean value, Set<Attribute> attributes) {
+    private LiteralBooleanSchema(boolean value, Set<Attribute> attributes, String message) {
         super(attributes);
         this.value = value;
+        this.message = message;
     }
 
     @Override
     protected String zodType() {
-        return String.format("literal(%s)", value);
+        String method = String.format("literal(%s)", value);
+        if (message != null) {
+            return MessageFormatter.addMessageToMethod(method, message);
+        }
+        return method;
+    }
+
+    public static LiteralBooleanSchema fromAttributes(Attribute booleanAttribute, Set<Attribute> attributes) {
+        if (booleanAttribute instanceof EqualsBooleanAttribute equalsBooleanAttribute) {
+            return new LiteralBooleanSchema(
+                    equalsBooleanAttribute.isValue(),
+                    Sets.difference(attributes, Set.of(equalsBooleanAttribute)),
+                    null);
+        }
+        if (booleanAttribute instanceof AttributeWithMessage attributeWithMessage) {
+            return new LiteralBooleanSchema(
+                    ((EqualsBooleanAttribute) attributeWithMessage.getAttribute()).isValue(),
+                    Sets.difference(attributes, Set.of(attributeWithMessage)),
+                    attributeWithMessage.getMessage());
+        }
+        throw new IllegalArgumentException("Unknown boolean attribute type " + booleanAttribute);
     }
 }

--- a/java-to-zod-core/src/main/java/sh/ivan/zod/schema/MessageFormatter.java
+++ b/java-to-zod-core/src/main/java/sh/ivan/zod/schema/MessageFormatter.java
@@ -1,0 +1,14 @@
+package sh.ivan.zod.schema;
+
+public final class MessageFormatter {
+    private MessageFormatter() {}
+
+    public static String addMessageToMethod(String method, String message) {
+        return method.replaceFirst("([^(])\\)$", "$1, { message: " + quoteMessage(message) + " })")
+                .replaceFirst("\\(\\)$", "({ message: " + quoteMessage(message) + " })");
+    }
+
+    private static String quoteMessage(String message) {
+        return "'" + message.replace("\\", "\\\\\\\\").replace("'", "\\\\'") + "'";
+    }
+}

--- a/java-to-zod-core/src/main/java/sh/ivan/zod/schema/attribute/Attribute.java
+++ b/java-to-zod-core/src/main/java/sh/ivan/zod/schema/attribute/Attribute.java
@@ -37,4 +37,8 @@ public interface Attribute {
                 .map(method -> "." + method)
                 .collect(Collectors.joining());
     }
+
+    default boolean isOfType(Class<? extends Attribute> clazz) {
+        return clazz.isAssignableFrom(this.getClass());
+    }
 }

--- a/java-to-zod-core/src/main/java/sh/ivan/zod/schema/attribute/AttributeWithMessage.java
+++ b/java-to-zod-core/src/main/java/sh/ivan/zod/schema/attribute/AttributeWithMessage.java
@@ -1,8 +1,11 @@
 package sh.ivan.zod.schema.attribute;
 
 import lombok.EqualsAndHashCode;
+import lombok.Getter;
 import lombok.ToString;
+import sh.ivan.zod.schema.MessageFormatter;
 
+@Getter
 @ToString
 @EqualsAndHashCode
 public class AttributeWithMessage implements Attribute {
@@ -21,13 +24,11 @@ public class AttributeWithMessage implements Attribute {
 
     @Override
     public String zodMethod() {
-        return attribute
-                .zodMethod()
-                .replaceFirst("([^(])\\)$", "$1, { message: " + quoteMessage() + " })")
-                .replaceFirst("\\(\\)$", "({ message: " + quoteMessage() + " })");
+        return MessageFormatter.addMessageToMethod(attribute.zodMethod(), message);
     }
 
-    private String quoteMessage() {
-        return "'" + message.replace("\\", "\\\\\\\\").replace("'", "\\\\'") + "'";
+    @Override
+    public boolean isOfType(Class<? extends Attribute> clazz) {
+        return attribute.isOfType(clazz);
     }
 }

--- a/java-to-zod-core/src/test/java/sh/ivan/zod/BooleanTest.java
+++ b/java-to-zod-core/src/test/java/sh/ivan/zod/BooleanTest.java
@@ -21,14 +21,40 @@ class BooleanTest extends JavaToZodConverterTest {
         var schema = converter.buildSchema(SucceededResult.class);
         assertThat(schema).isInstanceOf(ObjectSchema.class);
         var objectSchema = (ObjectSchema) schema;
-        assertThat(objectSchema.getFields()).hasSize(1);
+        assertThat(objectSchema.getFields()).hasSize(2);
         var failedSchema = objectSchema.getFields().get("failed");
         assertThat(failedSchema.asZodSchema()).isEqualTo("literal(false)");
+        var didFailSchema = objectSchema.getFields().get("didFail");
+        assertThat(didFailSchema.asZodSchema()).isEqualTo("literal(false).optional().nullable()");
     }
 
     static class SucceededResult {
         @AssertFalse
         public boolean failed;
+
+        @AssertFalse
+        public Boolean didFail;
+    }
+
+    @Test
+    void shouldHandleAssertFalseWithMessage() {
+        var schema = converter.buildSchema(SucceededResultWithMessage.class);
+        assertThat(schema).isInstanceOf(ObjectSchema.class);
+        var objectSchema = (ObjectSchema) schema;
+        assertThat(objectSchema.getFields()).hasSize(2);
+        var failedSchema = objectSchema.getFields().get("failed");
+        assertThat(failedSchema.asZodSchema()).isEqualTo("literal(false, { message: 'Must be false!' })");
+        var didFailSchema = objectSchema.getFields().get("didFail");
+        assertThat(didFailSchema.asZodSchema())
+                .isEqualTo("literal(false, { message: 'Must be false!' }).optional().nullable()");
+    }
+
+    static class SucceededResultWithMessage {
+        @AssertFalse(message = "Must be false!")
+        public boolean failed;
+
+        @AssertFalse(message = "Must be false!")
+        public Boolean didFail;
     }
 
     @Test
@@ -36,13 +62,39 @@ class BooleanTest extends JavaToZodConverterTest {
         var schema = converter.buildSchema(FailedResult.class);
         assertThat(schema).isInstanceOf(ObjectSchema.class);
         var objectSchema = (ObjectSchema) schema;
-        assertThat(objectSchema.getFields()).hasSize(1);
+        assertThat(objectSchema.getFields()).hasSize(2);
         var failedSchema = objectSchema.getFields().get("failed");
         assertThat(failedSchema.asZodSchema()).isEqualTo("literal(true)");
+        var didFailSchema = objectSchema.getFields().get("didFail");
+        assertThat(didFailSchema.asZodSchema()).isEqualTo("literal(true).optional().nullable()");
     }
 
     static class FailedResult {
         @AssertTrue
         public boolean failed;
+
+        @AssertTrue
+        public Boolean didFail;
+    }
+
+    @Test
+    void shouldHandleAssertTrueWithMessage() {
+        var schema = converter.buildSchema(FailedResultWithMessage.class);
+        assertThat(schema).isInstanceOf(ObjectSchema.class);
+        var objectSchema = (ObjectSchema) schema;
+        assertThat(objectSchema.getFields()).hasSize(2);
+        var failedSchema = objectSchema.getFields().get("failed");
+        assertThat(failedSchema.asZodSchema()).isEqualTo("literal(true, { message: 'Must be true!' })");
+        var didFailSchema = objectSchema.getFields().get("didFail");
+        assertThat(didFailSchema.asZodSchema())
+                .isEqualTo("literal(true, { message: 'Must be true!' }).optional().nullable()");
+    }
+
+    static class FailedResultWithMessage {
+        @AssertTrue(message = "Must be true!")
+        public boolean failed;
+
+        @AssertTrue(message = "Must be true!")
+        public Boolean didFail;
     }
 }

--- a/java-to-zod-maven-plugin-test/js/boolean-schema.test.js
+++ b/java-to-zod-maven-plugin-test/js/boolean-schema.test.js
@@ -1,34 +1,39 @@
-import { BooleanHolderSchema } from './schemas';
+import { BooleanHolderSchema, BooleanHolderWithMessagesSchema } from './schemas';
 import { isValid } from './test-utils';
 
-describe('BooleanHolderSchema', () => {
-    test('should support AssertTrue on boxed', () => {
-        const mustBeTrue = BooleanHolderSchema.shape.mustBeTrue;
-        expect(isValid(mustBeTrue, true)).toBe(true);
-        expect(isValid(mustBeTrue, null)).toBe(true);
-        expect(isValid(mustBeTrue, false)).toBe(false);
-    });
+const testSchema = (name, booleanSchema) => {
+    describe(name, () => {
+        test('should support AssertTrue on boxed', () => {
+            const mustBeTrue = booleanSchema.shape.mustBeTrue;
+            expect(isValid(mustBeTrue, true)).toBe(true);
+            expect(isValid(mustBeTrue, null)).toBe(true);
+            expect(isValid(mustBeTrue, false)).toBe(false);
+        });
 
-    test('should support AssertFalse on primitive', () => {
-        const mustBeFalse = BooleanHolderSchema.shape.mustBeFalse;
-        expect(isValid(mustBeFalse, false)).toBe(true);
-        expect(isValid(mustBeFalse, null)).toBe(false);
-        expect(isValid(mustBeFalse, true)).toBe(false);
-    });
+        test('should support AssertFalse on primitive', () => {
+            const mustBeFalse = booleanSchema.shape.mustBeFalse;
+            expect(isValid(mustBeFalse, false)).toBe(true);
+            expect(isValid(mustBeFalse, null)).toBe(false);
+            expect(isValid(mustBeFalse, true)).toBe(false);
+        });
 
-    test('should support boxed', () => {
-        const boxed = BooleanHolderSchema.shape.boxed;
-        expect(isValid(boxed, false)).toBe(true);
-        expect(isValid(boxed, true)).toBe(true);
-        expect(isValid(boxed, null)).toBe(true);
-        expect(isValid(boxed, 'foo')).toBe(false);
-    });
+        test('should support boxed', () => {
+            const boxed = booleanSchema.shape.boxed;
+            expect(isValid(boxed, false)).toBe(true);
+            expect(isValid(boxed, true)).toBe(true);
+            expect(isValid(boxed, null)).toBe(true);
+            expect(isValid(boxed, 'foo')).toBe(false);
+        });
 
-    test('should support primitive', () => {
-        const primitive = BooleanHolderSchema.shape.primitive;
-        expect(isValid(primitive, false)).toBe(true);
-        expect(isValid(primitive, true)).toBe(true);
-        expect(isValid(primitive, null)).toBe(false);
-        expect(isValid(primitive, 'foo')).toBe(false);
+        test('should support primitive', () => {
+            const primitive = booleanSchema.shape.primitive;
+            expect(isValid(primitive, false)).toBe(true);
+            expect(isValid(primitive, true)).toBe(true);
+            expect(isValid(primitive, null)).toBe(false);
+            expect(isValid(primitive, 'foo')).toBe(false);
+        });
     });
-});
+};
+
+testSchema('BooleanHolderSchema', BooleanHolderSchema);
+testSchema('BooleanHolderWithMessagesSchema', BooleanHolderWithMessagesSchema);

--- a/java-to-zod-maven-plugin-test/js/schemas.js
+++ b/java-to-zod-maven-plugin-test/js/schemas.js
@@ -12,6 +12,13 @@ export const BooleanHolderSchema = zod.object({
   primitive: zod.boolean(),
 });
 
+export const BooleanHolderWithMessagesSchema = zod.object({
+  mustBeTrue: zod.literal(true, { message: 'Needs to be true!' }).optional().nullable(),
+  mustBeFalse: zod.literal(false, { message: 'Needs to be false!' }),
+  boxed: zod.boolean().optional().nullable(),
+  primitive: zod.boolean(),
+});
+
 export const NumberHolderSchema = zod.object({
   maxed: zod.number().int().max(300).optional().nullable(),
   minned: zod.number().int().min(100),

--- a/java-to-zod-maven-plugin-test/src/main/java/sh/ivan/pojo/BooleanHolderWithMessages.java
+++ b/java-to-zod-maven-plugin-test/src/main/java/sh/ivan/pojo/BooleanHolderWithMessages.java
@@ -1,0 +1,10 @@
+package sh.ivan.pojo;
+
+import jakarta.validation.constraints.AssertFalse;
+import jakarta.validation.constraints.AssertTrue;
+
+public record BooleanHolderWithMessages(
+        @AssertTrue(message = "Needs to be true!") Boolean mustBeTrue,
+        @AssertFalse(message = "Needs to be false!") boolean mustBeFalse,
+        Boolean boxed,
+        boolean primitive) {}

